File: src/main/java/net/finance/specifications/ExpensesSpec.java
package net.finance.specifications;

import net.finance.entity.Category;
import net.finance.entity.Expense;
import net.finance.entity.User;
import org.springframework.data.jpa.domain.Specification;

import java.time.LocalDateTime;

public class ExpensesSpec {

    public static Specification<Expense> nameEquals(final String name) {
        return (expense, query, builder) -> name == null ? builder.conjunction() : builder.equal(expense.get("name"), name);
    }

    public static Specification<Expense> expireAtBetween(final LocalDateTime startDate, final LocalDateTime endDate) {
        return (expense, query, builder) -> startDate == null || endDate == null ? builder.conjunction() : builder.between(expense.get("expireAt"),
          startDate, endDate);
    }

    public static Specification<Expense> categoryEquals(final Category category) {
        return (expense, query, builder) -> category == null ? builder.conjunction() : builder.equal(expense.get("category"), category);
    }

    public static Specification<Expense> userEquals(final User user) {
        return (expense, query, builder) -> user == null ? builder.conjunction() : builder.equal(expense.get("user"), user);
    }

}

File: src/main/kotlin/com/finances/specifications/ExpensesSpec.kt
package com.finances.specifications

import com.finances.entity.Category
import com.finances.entity.Expense
import com.finances.entity.User
import org.springframework.data.jpa.domain.Specification
import java.time.LocalDateTime
import java.util.Optional
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.Root

object ExpensesSpec {

    @JvmStatic
    fun nameEquals(name: String?): Specification<Expense> {
        return Specification { expense: Root<Expense>, _, builder: CriteriaBuilder ->
            if (name.isNullOrBlank()) builder.conjunction()
            else builder.like(builder.lower(expense.get("name")), "%${name.toLowerCase()}%")
        }
    }

    @JvmStatic
    fun expireAtBetween(startDate: LocalDateTime?, endDate: LocalDateTime?): Specification<Expense> {
        return Specification { expense: Root<Expense>, _, builder: CriteriaBuilder ->
            if (startDate == null || endDate == null) builder.conjunction() else builder.between(
                expense.get("expireAt"),
                startDate,
                endDate
            )
        }
    }

    @JvmStatic
    fun categoryEquals(category: Optional<Category>): Specification<Expense> {
        return Specification { expense: Root<Expense>, _, builder: CriteriaBuilder ->
            if (category.isEmpty) builder.conjunction()
            else builder.equal(expense.get<Any>("category"), category.get())
        }
    }

    @JvmStatic
    fun userEquals(user: Optional<User>): Specification<Expense> {
        return Specification { expense: Root<Expense>, _, builder: CriteriaBuilder ->
            if (user.isEmpty) builder.conjunction()
            else builder.equal(expense.get<Any>("user"), user.get())
        }
    }
}
